src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
     return gens_stack
 
 
-def associate_generator_tables(gf, gen, gens):
-    """Associate the three tables needed to assign net gen to generators.
+def associate_generator_tables(
+    gens: pd.DataFrame, gf: pd.DataFrame, gen: pd.DataFrame, bf_by_gens: pd.DataFrame
+) -> pd.DataFrame:
+    """Associate the three tables needed to assign net gen and fuel to generators.
+
+    The ``generation_fuel_eia923`` table's data is reported at the
+    :py:const:`IDX_PM_ESC` granularity. Each generator has one ``prime_mover_code``, but
+    potentially several ``energy_source_code``s (reported in as primary source,
+    secondary source, etc.). Much of the data in ``generation_fuel_eia923`` not
+    associated with primary energy sources, so before merging the data with the
+    generators, we employ :func:``stack_generators``.
+
+    After the generator's energy source codes have been stacked, we group the other
+    data tables and merge them with the generators. We add suffixes to the data columns
+    to identify the source table.
+
+    This function also removes inactive generators so that we don't associate any net
+    generation or fuel to those generators. See :func:`remove_inactive_generators` for
+    more details.
+
+    There are some records in the data tables that have primary key values that do no
+    match up with any of those values in the generators table. We employ
+    :func:`_allocate_unassociated_records` to make sure those records are associated.
```

Comment:
This is weird and complicating. Can you add a simple example (is this like, a PM that doesn't match up, or a fuel type that appears in the `gf` table but not in the list of `energy_source_code`s associated with the generator? Can we note that this is an internal inconsistency in the reported data?